python/lsst/pipe/base/tests/pipelineStepTester.py

```diff
@@ -0,0 +1,118 @@
+# This file is part of pipe_base.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Utility to facilitate testing of pipelines consisting of multiple steps.
+"""
+
+__all__ = ["PipelineStepTester"]
+
+import dataclasses
+import unittest
+
+from lsst.daf.butler import Butler, DatasetType
+from lsst.pipe.base import Pipeline, PipelineDatasetTypes
+
+
+@dataclasses.dataclass
+class PipelineStepTester:
+    """Utility class which facilitates testing of pipelines, optionally
+    consisting of multiple steps.
+
+    Two sets will be constructed by looping over the entire pipeline or all
+    named subsets within the pipeline: `pure_inputs` and `all_outputs`.
+
+    The `pure_inputs` set consists of all inputs which must be constructed and
+    provided as an input into the pipeline (i.e., they will not be generated
+    by the named pipeline).
+
+    The `all_outputs` set consists of all dataset types which are generated by
+    the named pipeline, either as intermediates or as final outputs.
+
+    These sets will be checked against user-supplied sets to ensure that the
+    named pipeline may still be run without raising a missing data error.
+
+    Parameters
+    ----------
+    filename : `str`
+        The full path to the pipeline YAML
+    step_suffixes : `list` of `str`
```

Comment:
For documentation of containers, you should use this style:
https://developer.lsst.io/python/numpydoc.html#sequence-types

Reply:
Why are these lists, as opposed to say sets? Does the order matter here?

Reply:
Documentation updated as per the referenced style guide.

Reply:
Yes, the order is important in `step_suffixes`. Specifically, the pipeline needs to be progressively built in the order in which it's executed. This allows us to build up a list of overall expected input dataset types, and known expected outputs.

The `initial_dataset_types` parameter needs to be a list as it contains unhashable types.